### PR TITLE
Reimplement getRequiredPlugins for go sdk

### PR DESCRIPTION
**Pull request by EvanBoyle**
This uses `go list -m -json all`. We must use -m `module` mode as this is the only way to get the version info as far as I can tell. This means that this approach only works for users using modules, not glide, dep, etc. This is the direction the community is going, and the direction we've moved our docs/examples so this should be fine. But as a result, I've made the implementation lenient. It doesn't fail if we run into errors along the way, just logs some messages saying we failed to acquire plugins. This is strictly better than what we have today, and we can always revisit this decision in the future as the ecosystem evolves. 

```
Evans-MBP:aws-go-webserver evanboyle$ pulumi preview
Previewing update (dcgo):
[resource plugin aws-1.28.0] installing
Downloading plugin: 64.58 MiB / 64.58 MiB [=========================] 100.00% 7s
Moving plugin... done.
     Type                      Name               Plan       
 +   pulumi:pulumi:Stack       go-webserver-dcgo  create     
 +   ├─ aws:ec2:SecurityGroup  web-secgrp         create     
 +   └─ aws:ec2:Instance       web-server-www     create     
 
Resources:
    + 3 to create

Permalink: https://app.pulumi.com/EvanBoyle/go-webserver/dcgo/previews/a761b512-4136-4caa-b504-6843e72c0193

```

Fixes https://github.com/pulumi/pulumi/issues/1549 
Fixes https://github.com/pulumi/pulumi/issues/4211

### CODE REVIEW

File: sdk/go/pulumi-language-go/main.go [hunk 1]
```diff
@@ -157,10 +160,71 @@ func newLanguageHost(engineAddress, tracing string) pulumirpc.LanguageRuntimeSer
...
+// with respect to plugin acquisition
+type modInfo struct {
+	Path    string
+	Version string
+}
+
+func getPluginName(modName string) string {
+	// github.com/pulumi/pulumi-aws/sdk/... => aws
+	pluginPart := strings.Split(modName, "/")[2]
+	return strings.SplitN(pluginPart, "-", 2)[1]
```
[the remark this answers is not in the document]
Comment (lukehoban) [reply sdk/go/pulumi-language-go/main.go:173]: Oh - I see that the caller ensures they can't.

File: sdk/go/pulumi-language-go/main.go [hunk 2]
```diff
@@ -157,10 +160,71 @@ func newLanguageHost(engineAddress, tracing string) pulumirpc.LanguageRuntimeSer
...
+}
+
 // GetRequiredPlugins computes the complete set of anticipated plugins required by a program.
+// We're lenient here as this relies on the `go list` command and the use of modules.
+// If the consumer insists on using some other form of dependency management tool like
+// dep or glide, the list command fails with "go list -m: not using modules"
 func (host *goLanguageHost) GetRequiredPlugins(ctx context.Context,
```
Comment (lukehoban) [inline sdk/go/pulumi-language-go/main.go:180]: Can we add any test coverage of this?

File: sdk/go/pulumi-language-go/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (EvanBoyle) [reply sdk/go/pulumi-language-go/main.go]: Correct. I refactored this to make the relationship clear in case we change this code in the future.

File: sdk/go/pulumi-language-go/main.go [hunk 3]
```diff
@@ -157,10 +160,71 @@ func newLanguageHost(engineAddress, tracing string) pulumirpc.LanguageRuntimeSer
...
+				break
+			}
+			logging.V(5).Infof("GetRequiredPlugins: Error parsing list output: %s", err.Error())
+			return &pulumirpc.GetRequiredPluginsResponse{}, nil
+		}
+
+		if strings.HasPrefix(m.Path, "github.com/pulumi/pulumi-") {
+			plugin := &pulumirpc.PluginDependency{
+				Name:    getPluginName(m.Path),
+				Version: m.Version,
```
[the remark this answers is not in the document]
Comment (EvanBoyle) [reply sdk/go/pulumi-language-go/main.go]: Good point. I updated this to handle non-tag commits (we normalize the patch to `0` to be compatible).  The only way we can get something like `v0.0.0` is when the module is new and has no associated tags, which should never happen now that we've upgraded to the new module structure. I suppose we could filter out 0.0.0 if we want to be extra careful. I'm open to suggestions here.

File: sdk/go/pulumi-language-go/main.go (hunk 2 above)
Comment (EvanBoyle) [reply sdk/go/pulumi-language-go/main.go:203]: I refactored to test the code to test the filtering and conversion of the module result from `go list` into a plugin description. 

This seems sufficient to me. Any further validation would require an integration test with a dependency on pulumi-aws.

File: sdk/go/pulumi-language-go/main.go [hunk 4]
```diff
@@ -157,10 +160,71 @@ func newLanguageHost(engineAddress, tracing string) pulumirpc.LanguageRuntimeSer
...
+
+	logging.V(5).Infof("GetRequiredPlugins: Determining pulumi packages")
+
+	gobin, err := findExecutable("go")
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't find go binary")
+	}
+
+	// don't wire up stderr so non-module users don't see error output from list
+	cmd := exec.Command(gobin, "list", "-m", "-json", "all")
```
[the remark this answers is not in the document]
Comment (EvanBoyle) [reply sdk/go/pulumi-language-go/main.go:214]: Yup, tried this initially and learned about the internal restriction.

File: sdk/go/pulumi-language-go/main.go [hunk 5]
```diff
@@ -157,10 +161,89 @@ func newLanguageHost(engineAddress, tracing string) pulumirpc.LanguageRuntimeSer
...
+	// pinning to a specific commit can result in a minor version that doesn't exist
+	// v1.29.1-0.20200403140640-efb5e2a48a86 (first commit after 1.29.0 release)
+	version := fmt.Sprintf("v%v.%v.%v", v.Major, v.Minor, 0)
+
+	plugin := &pulumirpc.PluginDependency{
+		Name:    name,
+		Version: version,
+		Kind:    "resource",
+	}
+	logging.V(5).Infof("GetRequiredPlugins: Found plugin name: %s, version: %s", plugin.Name, plugin.Version)
```
Comment (lukehoban) [inline sdk/go/pulumi-language-go/main.go:194]: This is duplicated and can be removed here.

File: sdk/go/pulumi-language-go/main.go [hunk 6]
```diff
@@ -157,10 +161,89 @@ func newLanguageHost(engineAddress, tracing string) pulumirpc.LanguageRuntimeSer
...
+		if err := dec.Decode(&m); err != nil {
+			if err == io.EOF {
+				break
+			}
+			logging.V(5).Infof("GetRequiredPlugins: Error parsing list output: %s", err.Error())
+			return &pulumirpc.GetRequiredPluginsResponse{}, nil
+		}
+
+		plugin, err := m.getPlugin()
+		if err == nil {
```
Comment (lukehoban) [inline sdk/go/pulumi-language-go/main.go:238]: Should we log in the err case?

File: sdk/go/pulumi-language-go/main.go [hunk 7]
```diff
@@ -157,10 +161,89 @@ func newLanguageHost(engineAddress, tracing string) pulumirpc.LanguageRuntimeSer
...
+	name := strings.SplitN(pluginPart, "-", 2)[1]
+
+	v, err := semver.ParseTolerant(m.Version)
+	if err != nil {
+		return nil, errors.New("module does not have semver compatible version")
+	}
+	// downgrade to zero for the patch
+	// pinning to a specific commit can result in a minor version that doesn't exist
+	// v1.29.1-0.20200403140640-efb5e2a48a86 (first commit after 1.29.0 release)
+	version := fmt.Sprintf("v%v.%v.%v", v.Major, v.Minor, 0)
```
Comment (lukehoban) [inline sdk/go/pulumi-language-go/main.go:187]: In the case that a precise version was specified - this seems wrong. 

Also in general - I’m not positive this won’t work - for Kubernetes for example I think dev builds are prerelease for major version bumps?  And is dependent on some very specific things about how we currently version dev tags.  Anything more reliable we could do here?  

Perhaps no and this is still the best we can do heuristically and is worth it?

Comment (EvanBoyle) [reply sdk/go/pulumi-language-go/main.go:187]: Good point. I'll lift some of the pseduo version parsing code out of internal to make this more robust.

**Comment by EvanBoyle** [comment]
@lukehoban Please take another look. I took some code out of internal that detects pseudo versions so we can special case their handling. This allows us to leave patch versions alone in the normal case where our version matches a tag. This also allows us to handle beta/dev tags. I added more unit test cases to document the behavior of more of these corner cases. 

I manually tested the `v2.0.0-beta.1` release of AWS:

```
Evans-MBP:aws-go-webserver evanboyle$ pulumi preview
Previewing update (dcgo):
[resource plugin aws-2.0.0-beta.1] installing
Downloading plugin: 64.56 MiB / 64.56 MiB [=========================] 100.00% 3s
Moving plugin... done.
     Type                      Name               Plan       Info
 +   pulumi:pulumi:Stack       go-webserver-dcgo  create     2 messages
 +   ├─ aws:ec2:SecurityGroup  web-secgrp         create     
 +   └─ aws:ec2:Instance       web-server-www     create     
 
Diagnostics:
  pulumi:pulumi:Stack (go-webserver-dcgo):
    go: downloading github.com/pulumi/pulumi-aws/sdk/v2 v2.0.0-beta.1
    go: extracting github.com/pulumi/pulumi-aws/sdk/v2 v2.0.0-beta.1
 

Permalink: https://app.pulumi.com/EvanBoyle/go-webserver/dcgo/previews/47722fda-8bc9-48d6-8776-ba2c3f4e7558
```

File: sdk/go/common/util/buildutil/semver_test.go [hunk 8]
```diff
@@ -36,3 +36,14 @@ func TestVersions(t *testing.T) {
...
 }
+
+func TestPseduoVersion(t *testing.T) {
+	pseudoVersion := "v1.29.1-0.20200403140640-efb5e2a48a86"
+	assert.True(t, IsPseudoVersion(pseudoVersion))
+
+	tagVersion := "v1.29.0"
+	assert.False(t, IsPseudoVersion(tagVersion))
+
+	betaVersion := "v.1.29.0-beta.1"
```
Comment (lukehoban) [inline sdk/go/common/util/buildutil/semver_test.go:47]:
```suggestion
	betaVersion := "v1.29.0-beta.1"
```

?
